lib/srv/server/ssm_install.go

```diff
@@ -0,0 +1,150 @@
+/*
+Copyright 2022 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package server
+
+import (
+	"context"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/aws/request"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/aws/aws-sdk-go/service/ssm"
+	"github.com/aws/aws-sdk-go/service/ssm/ssmiface"
+	"github.com/gravitational/teleport/api/types/events"
+	apievents "github.com/gravitational/teleport/api/types/events"
+	libevents "github.com/gravitational/teleport/lib/events"
+	"github.com/gravitational/trace"
+	"golang.org/x/sync/errgroup"
+)
+
+// SSMInstallerConfig represents configuration for an SSM install
+// script executor.
+type SSMInstallerConfig struct {
+	// Emitter is an events emitter.
+	Emitter apievents.Emitter
+}
+
+// SSMInstaller handles running SSM commands that install Teleport on EC2 instances.
+type SSMInstaller struct {
+	SSMInstallerConfig
+}
+
+type SSMRunRequest struct {
+	// DocumentName is the name of the SSM document to run.
+	DocumentName string
+	// SSM is an SSM API client.
+	SSM ssmiface.SSMAPI
+	// Instances is the list of instances that will have the SSM
+	// document executed on them.
+	Instances []*ec2.Instance
+	// Params is a list of parameters to include when executing the
+	// SSM document.
+	Params map[string]string
+	// Region is the region instances are present in, used in audit
+	// events.
+	Region string
+	// AccountID is the AWS account being used to execute the SSM document.
+	AccountID string
+}
+
+// NewSSMInstaller returns a new instance of the SSM installer that installs Teleport on EC2 instances.
+func NewSSMInstaller(cfg SSMInstallerConfig) *SSMInstaller {
+	return &SSMInstaller{
+		SSMInstallerConfig: cfg,
+	}
+}
+
+// Run executes the SSM document and then blocks until the command has completed.
+func (i *SSMInstaller) Run(ctx context.Context, req SSMRunRequest) error {
```

Comment:
check my comment above about multiple runs for the same machine, but if possible should we just store the ids of the ec2 machines which we run SSM against but didn't finish yet?